Adds support for base-4.12.0.0

jbrechtel opened the pull request.

ysangkok commented on src/Flipstone/Prelude.hs:
```diff
@@ -69,7 +70,11 @@ module Flipstone.Prelude
  , Semigroup((<>), sconcat, stimes)
  , Monoid(mempty, mconcat) -- `mappend` is redundant and not part of the minimal complete definition
  , Monad((>>=), (>>)) -- `return` is redundant and not part of the minimal complete definition
+#if MIN_VERSION_base(4,13,0)
```
Why 4.13? MonadFail exists in 4.12.

But there is an even more extreme strategy: Monad is in Prelude, we are making a replacement Prelude. So why not have our Prelude enforce the MonadFail migration for *all* GHC versions since 8.0, when the MonadFail transitional strategy started? We don't need to support 7.10 anyway. Even if fail is in Monad up to 8.8, we don't necessarily need to treat it like it is, since we get the right to redefine the Prelude in this package.

qxjit replied:
@ysangkok @jbrechtel I would certainly vote for using the `MonadFail(fail)` import in the earliest version that it's valid to do so (4.9?)

And if we don't need to support anything earlier than that version we can remove the conditional compilation entirely

ysangkok replied:
Right, 4.9 sounds looking at [base version correspondence list](https://wiki.haskell.org/Base_package) and the [MonadFail wiki page](https://wiki.haskell.org/MonadFail_Proposal).

BTW, my proposal about backporting the whole MonadFail proposal to all releases won't work because Monad is hardwired and even people who use custom Preludes expect the Monad they get there to be the same as the one all their libraries refer to.

qxjit replied:
Oh yes, I wasn't imagine backporting anything. Just exporting `fail` from `MonadFail` rather than the one from `Monad`, as far back as is possible.

jbrechtel replied:
Oops, yeah, wrong bounds my bad.

jbrechtel commented:
I updated this to use `MonadFail` directly from `Control.Monad.Fail` so that ifdef is gone and then I fixed the `Data.Foldable` import to keep the `Foldable` typeclass and only drop `foldMap'` when we're below base 4.13

ysangkok reviewed (changes requested):
@jbrechtel I noticed that the common elements are listed twice for each branch of the ifdef.

So I tried to fix it with 

```
#if MIN_VERSION_base(4,13,0)                                                        
   , Foldable(foldMap')                                                               
#endif                                                                              
   , Foldable(fold, foldMap, foldr, foldr', foldl', toList, null, length, elem, maximum, sum, product)
```

but that results in a duplicate export.  However, there is another way:

```
  , Foldable(                                                                    
#if MIN_VERSION_base(4,13,0)                                                    
       foldMap',                                                                  
#endif                                                                          
       fold, foldMap, foldr, foldr', foldl', toList, null, length, elem, maximum, sum, product
    )
```

What do you think?

jbrechtel commented:
I like it. PR updated.

ysangkok reviewed (changes requested):
The same trick can be applied to the `import` ;)
